core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -105,65 +168,255 @@
+            (BeforeRollbackTransactionListener.class).values();
+        for (BeforeRollbackTransactionListener listener : listeners) {
+            try {
+                listener.beforeRollback(status);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" beforeRollback callback failed: "+ t);
+                beforeRollbackFailure = t;
+            }
+        }
+        if (beforeRollbackFailure != null){
+            throw beforeRollbackFailure;
+        }
+    }
+
+    /**
+     * Safely perform {@link PlatformTransactionManager#rollback(TransactionStatus)}.
+     *
+     * This method is responsible for safely logging any notification failures as warnings, rather than failing outright.
+     *
+     * @param context context used to obtain listeners to notify
+     * @param transactionManager manager responsible for execution
+     * @param status transaction status
+     */
     private static void doRollback(ApplicationContext context,
                                    PlatformTransactionManager transactionManager,
-                                   @Nullable TransactionStatus transaction) {
+                                   @Nullable TransactionStatus status) {
+        if ( status == null || status.isCompleted()){
+            return; // nothing to do transaction already completed
+        }
         try {
-            if (transaction != null && !transaction.isCompleted()) {
+            fireBeforeRollback( context, transactionManager, status);
+        }
+        catch (Throwable t) {
+            Log.warning(Log.JEEVES, "Rolling back transaction - before:", t);
+        }
 
-                try {
-                    Collection<BeforeRollbackTransactionListener> listeners = context.getBeansOfType
-                        (BeforeRollbackTransactionListener.class).values();
-                    for (BeforeRollbackTransactionListener listener : listeners) {
-                        listener.beforeRollback(transaction);
-                    }
-                } finally {
-                    transactionManager.rollback(transaction);
-                    Collection<AfterRollbackTransactionListener> listeners = context.getBeansOfType(
-                        AfterRollbackTransactionListener.class).values();
-                    for (AfterRollbackTransactionListener listener : listeners) {
-                        listener.afterRollback(transaction);
-                    }
-                }
-            }
-            //what if the transaction is completed?
-            //maybe then we shouldn't be here
-        } catch (Throwable t) {
+        try {
+            transactionManager.rollback(status);
+        }
+        catch (Throwable t) {
             Log.error(Log.JEEVES, "ERROR rolling back transaction", t);
         }
+        finally {
+            try {
+                fireAfterRollback(context, transactionManager, status);
```

Comment:
In the original code if `transactionManager.rollback` fails, the after commit listeners were not executed, with this change seem are executed.

Not very clear if it's relevant to execute them, unless they could handle the transaction rollback failure?

Reply:
I updated based on your feedback, the listeners can tell if the rollback completed (or failed). I did make an explicit log message so we can see what is going on.